src/execnet/gateway_base.py

```diff
@@ -376,6 +386,19 @@ def _try_send_to_primary_thread(self, reply):
                 # wake up primary thread
                 primary_thread_task_ready.set()
                 return True
+            elif (
+                self.execmodel.backend == "main_thread_only"
+                and self._primary_thread_task is not None
+            ):
+                self._primary_thread_task.waitfinish()
+                self._primary_thread_task = reply
+                # wake up primary thread (it's okay if this is already set
+                # because we waited for the previous task to finish above
+                # and integrate_as_primary_thread will not clear it when
+                # it enters self._running_lock if it detects that a new
+                # task is available)
+                primary_thread_task_ready.set()
+                return True
```

Comment:
I suppose that since we have the [deadlock error handling in _local_schedulexec](https://github.com/pytest-dev/execnet/pull/243#discussion_r1493856210)  we could possibly remove support for waiting for the previous task in  _try_send_to_primary_thread, but it could still be useful if there's ever a context where one would want to submit multiple spawn calls and have them all run serially in the main thread, with spawn blocking to provide back pressure. I audited the internal spawn calls and it seems like I've found one such context in the `safe_terminate` function where it calls spawn in a loop, but actually it instantiates the WorkerPool with default `hasprimary=False` which disables main thread usage:
```python
def safe_terminate(execmodel, timeout, list_of_paired_functions):
    workerpool = WorkerPool(execmodel)

    def termkill(termfunc, killfunc):
        termreply = workerpool.spawn(termfunc)
        try:
            termreply.get(timeout=timeout)
        except OSError:
            killfunc()

    replylist = []
    for termfunc, killfunc in list_of_paired_functions:
        reply = workerpool.spawn(termkill, termfunc, killfunc)
        replylist.append(reply)
    for reply in replylist:
        reply.get()
    workerpool.waitall(timeout=timeout)
```
Meanwhile, I noticed that [related test cases have incorrect conditions for pytest.xfail calls](https://github.com/pytest-dev/execnet/pull/243/files#r1494044060).

Reply:
Elimination of [false-positive deadlock errors](https://github.com/pytest-dev/execnet/pull/243#discussion_r1495103598) now requires support for blocking in _try_send_to_primary_thread, but usage of the WorkerGateway _executetask_complete event guarantees that this blocking will never lead to a deadlock.